stl/inc/semaphore

```diff
@@ -0,0 +1,254 @@
+        }
+        _STL_VERIFY(_Current >= 0 && _Current <= _Least_max_value,
+            "Semaphore counter shall be non-negative and shall not overflow max()");
+        return _Counter.compare_exchange_weak(
+            _Current, _Current - 1, std::memory_order_acquire, std::memory_order_relaxed);
+    }
+
+    template <class _Rep, class _Period>
+    [[nodiscard]] bool try_acquire_for(const chrono::duration<_Rep, _Period>& _Rel_time) {
+        auto _Rel_time_ms  = chrono::duration_cast<chrono::duration<unsigned long, milli>>(_Rel_time);
+        auto _Deadline     = __std_atomic_wait_get_deadline(_Rel_time_ms.count());
+        ptrdiff_t _Current = _Counter.load(memory_order_relaxed);
+        for (;;) {
+            while (_Current == 0) {
+                const auto _Remaining_timeout = __std_atomic_wait_get_remaining_timeout(_Deadline);
+                if (_Remaining_timeout == 0) {
+                    return false;
+                }
+                __std_atomic_wait_direct(&_Counter, &_Current, sizeof(_Current), _Remaining_timeout);
+                _Current = _Counter.load(memory_order_relaxed);
+            }
+            _STL_VERIFY(_Current >= 0 && _Current <= _Least_max_value,
+                "Semaphore counter shall be non-negative and shall not overflow max()");
+            if (_Counter.compare_exchange_weak(
+                    _Current, _Current - 1, std::memory_order_acquire, std::memory_order_relaxed)) {
+                return true;
+            }
+        }
+    }
+
+    template <class _Clock, class _Duration>
+    [[nodiscard]] bool try_acquire_until(const chrono::time_point<_Clock, _Duration>& _Abs_time) {
+        ptrdiff_t _Current = _Counter.load(memory_order_relaxed);
+        for (;;) {
+            while (_Current == 0) {
+                auto _Now = _Clock::now();
+                if (_Now >= _Abs_time) {
+                    return false;
+                }
+
+                auto _Rel_time = _Abs_time - _Now;
+                chrono::duration<unsigned long, milli> _Rel_time_ms;
+                static constexpr auto _Ten_days = chrono::hours(24 * 10);
+                if (_Rel_time >= _Ten_days) {
+                    _Rel_time_ms = _Ten_days;
+                } else {
+                    _Rel_time_ms = chrono::duration_cast<chrono::duration<unsigned long, milli>>(_Rel_time);
+                }
+                __std_atomic_wait_direct(&_Counter, &_Current, sizeof(_Current), _Rel_time_ms.count());
+                _Current = _Counter.load(memory_order_relaxed);
+            }
+            _STL_VERIFY(_Current >= 0 && _Current <= _Least_max_value,
+                "Semaphore counter shall be non-negative and shall not overflow max()");
+            if (_Counter.compare_exchange_weak(
+                    _Current, _Current - 1, std::memory_order_acquire, std::memory_order_relaxed)) {
+                return true;
+            }
+        }
+    }
+
+private:
+    atomic<ptrdiff_t> _Counter;

```

Comment:
Rather than always using `ptrdiff_t` maybe choose type based on `_Least_max_value`?

Reply:
I was thinking about it, but I'm not sure if it worth doing. 
Have to add casts and meta-programming for apparently a very minor size saving (which is likely to be ultimately wasted, normally you'd want these things on separate cache line)

Reply:
By the way. I'm going to add atomic wait to `atomic_shared_ptr` soon too. (Or did you want to do it yourself?)

Reply:
> I was thinking about it, but I'm not sure if it worth doing.
> Have to add casts and meta-programming for apparently a very minor size saving (which is likely to be ultimately wasted, normally you'd want these things on separate cache line)

The size improvement is minor but someone might find it useful. The meta-programming cost is not big.

> By the way. I'm going to add atomic wait to `atomic_shared_ptr` soon too. (Or did you want to do it yourself?)

I'm not an `atomic` expert (especially with the waiting stuff) and I don't have much time for contributions lately but I will try to leave some code review from time to time.

Reply:
I implemented this. `int8_t` -> `int32_t` -> `ptrdiff_t`.
I've skipped `int16_t` for now. 

My understanding is what we should optimize more here is code size, not variable size. For code size on x64, changing from 64-bit to 32-bit is a win, but changing from 32-bit to 16-bit is a loss, changing from 32-bit to 8-bit is a loss too, but a very minor.

I've also made `max()` reflecting maximum type of variable rather than keeping template parameter to simplify checks.

I've also made default template parameter to `INT_MAX` to actually benefit from this by default.

Reply:
Enabled for 16-bytes type too, apparently disadvantage is not that big.

It is still open question isn't it harmful on arm though.

Reply:
I reverted this optimization to implement another optimization to avoid waking too many waiting threads.

(However if I'll have to revert the new optimization, I put back the object space optimization)